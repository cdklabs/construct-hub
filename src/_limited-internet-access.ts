import { createHash } from 'crypto';
import * as fs from 'fs';
import { resolve, join } from 'path';
import {
  IVpc,
  ISecurityGroup,
  SecurityGroup,
  Port,
  Peer,
  IPeer,
  Connections,
  CfnPrefixList,
} from '@aws-cdk/aws-ec2';
import { Construct, Tags } from '@aws-cdk/core';
import { S3PrefixList } from './s3';

/**
 * Creates SecurityGroups where "sensitive" operations should be listed,
 * which only allows DNS requests to be issued within the VPC (to the local
 * Route53 resolver), as well as HTTPS (port 443) traffic to:
 * - allow-listed IP ranges
 * - endpoints within the same SecurityGroup.
 *
 * This returns MULTIPLE security groups in order to avoid hitting the maximum
 * count of rules per security group, which is relatively low, and prefix
 * lists count as their expansions.
 *
 * There is also a limit of how many security groups can be bound to a network
 * interface (defaults to 5), so there is only so much we can do here.
 *
 * @param scope the scope in which to attach new constructs.
 * @param vpc the VPC in which a SecurityGroup is to be added.
 */
export function createRestrictedSecurityGroups(
  scope: Construct,
  vpc: IVpc
): ISecurityGroup[] {
  const securityGroups = new Array<ISecurityGroup>();

  securityGroups.push(createInternalTrafficSecurityGroup(scope, vpc));

  const ALLOW_LIST_DIR = resolve(
    __dirname,
    '..',
    'resources',
    'vpc-allow-lists'
  );
  for (const file of fs.readdirSync(ALLOW_LIST_DIR)) {
    const matches = /^(.+)-(IPv4|IPv6)\.txt$/.exec(file);
    if (matches == null) {
      throw new Error(
        `Allow-list file ${file} in ${ALLOW_LIST_DIR} is invalid: file name must end in IPv4.txt or IPv6.txt`
      );
    }
    const [, namespace, ipLabel] = matches;

    const entries = parsePrefixList(join(ALLOW_LIST_DIR, file));

    if (entries.length === 0) {
      continue;
    }

    // We use a SHA-1 digest of the list of prefixes to be sure we create a
    // whole new prefix list whenever it changes, so we are never bothered by
    // the maxEntries being what it is.
    const hash = entries
      .reduce((h, { cidr }) => h.update(cidr).update('\0'), createHash('SHA1'))
      .digest('hex');

    // Note - the `prefixListName` is NOT a physical ID, and so updating it
    // will NOT cause a replacement. Additionally, it needs not be unique in
    // any way.
    const pl = new CfnPrefixList(scope, `${namespace}.${ipLabel}#${hash}`, {
      addressFamily: ipLabel,
      prefixListName: `${namespace}.${ipLabel}`,
      entries,
      maxEntries: entries.length,
    });
    // Note: the CfnPrefixList above uses a `.` separator bewteen namespace and
    // ipLabel, since we use a `-` here, we are not colliding. The has is used
    // here to replace the SG when the PL is updated. Failure to do so may
    // result in the SG having both the old & updated PLs in its rule set until
    // the CloudFormation clean-up phase completes, which might exceed the
    // maximum amount of rules allowed on a SG.
    const descr = `${namespace}-${ipLabel}`;
    const sg = new SecurityGroup(scope, `${descr}#${hash}`, {
      allowAllOutbound: false,
      description: `${scope.node.path}/${descr}`,
      vpc,
    });

    // We intentionally ONLY allow HTTPS though there...
    sg.connections.allowTo(
      NamedPeer.from(Peer.prefixList(pl.attrPrefixListId), pl.node.path),
      Port.tcp(443),
      `to ${namespace} (${ipLabel})`
    );

    Tags.of(sg).add('Name', `${namespace}.${ipLabel}`);

    securityGroups.push(sg);
  }

  return securityGroups;
}

/**
 * Creates a SecurityGroup that allows traffic to flow freely between
 * endpoints within itself on port 443, to the local Route53 resolver on DNS
 * ports, and to the region's AW S3 prefix list on port 443.
 *
 * @param scope the scope in which to attach the new Security Group.
 * @param vpc the VPC in which the SecurityGroup will be created.
 */
function createInternalTrafficSecurityGroup(
  scope: Construct,
  vpc: IVpc
): ISecurityGroup {
  const sg = new SecurityGroup(scope, 'InternalTraffic', {
    allowAllOutbound: false,
    description: `${scope.node.path}/SG`,
    vpc,
  });

  // Allow all traffic within the security group on port 443
  sg.connections.allowInternally(
    Port.tcp(443),
    'Traffic within this SecurityGroup'
  );

  // Allow access to S3. This is needed for the S3 Gateway endpoint to work.
  sg.connections.allowTo(
    NamedPeer.from(
      Peer.prefixList(new S3PrefixList(scope, 'S3-PrefixList').prefixListId),
      'AWS S3'
    ),
    Port.tcp(443),
    'to AWS S3'
  );

  // Allow making DNS requests, there should be a Route53 resolver wihtin the VPC.
  sg.connections.allowTo(
    Peer.ipv4(vpc.vpcCidrBlock),
    Port.tcp(53),
    'to Route53 DNS resolver'
  );
  sg.connections.allowTo(
    Peer.ipv4(vpc.vpcCidrBlock),
    Port.udp(53),
    'to Route53 DNS resolver'
  );

  return sg;
}

/**
 * Parses the PrefixList in the designated path.
 *
 * @param filePath the file containing the prefix list.
 */
export function parsePrefixList(filePath: string): CidrBlock[] {
  return (
    fs
      .readFileSync(filePath, 'utf8')
      .split(/\n/)
      .map((line) => {
        const match = /^\s*([^\s]+)?\s*(?:#.*)?$/.exec(line);
        if (!match) {
          throw new Error(`Invalid line in allow list ${filePath}: ${line}`);
        }
        const [, cidr] = match;
        return cidr;
      })
      // Remove empty lines.
      .filter((cidr) => !!cidr)
      .sort()
      .map((cidr) => ({ cidr }))
  );
}

interface CidrBlock {
  readonly cidr: string;
}

/**
 * This is to work around an issue where the peer's `uniqueId` is a token for
 * our PrefixList values, and this causes the VPC construct to "de-duplicate"
 * all of them (it considers they are identical).
 *
 * There is a fix in the latest EC2 library, however that fix isn't great
 * either, as it addresses the problem at the wrong location (in the in/egress
 * rule, instead of in the peer).
 *
 * Basically, this ensures the `uniqueId` is some string we control, so we
 * remain faithful to the declaraiton intent.
 */
class NamedPeer implements IPeer {
  public static from(peer: IPeer, name: string) {
    return new NamedPeer(peer, name);
  }

  public readonly connections: Connections = new Connections({ peer: this });

  private constructor(
    private readonly peer: IPeer,
    public readonly uniqueId: string
  ) {}

  public get canInlineRule() {
    return this.peer.canInlineRule;
  }

  public toIngressRuleConfig() {
    return this.peer.toIngressRuleConfig();
  }

  public toEgressRuleConfig() {
    return this.peer.toEgressRuleConfig();
  }
}
