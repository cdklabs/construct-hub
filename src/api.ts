// this file includes types that are part of the library's public API

import * as certificatemanager from '@aws-cdk/aws-certificatemanager';
import { IAlarmAction } from '@aws-cdk/aws-cloudwatch';
import * as route53 from '@aws-cdk/aws-route53';

export * from './backend/deny-list/api';
export * from './backend/license-list/api';
export * from './codeartifact/api';
export * from './monitoring/api';
export * from './overview-dashboard/api';

/**
 * Domain configuration for the website.
 */
export interface Domain {
  /**
   * The root domain name where this instance of Construct Hub will be served.
   */
  readonly zone: route53.IHostedZone;

  /**
   * The certificate to use for serving the Construct Hub over a custom domain.
   *
   * @default - a DNS-Validated certificate will be provisioned using the
   *            provided `hostedZone`.
   */
  readonly cert: certificatemanager.ICertificate;

  /**
   * Whether the certificate should be monitored for expiration, meaning high
   * severity alarms will be raised if it is due to expire in less than 45 days.
   *
   * @default true
   */
  readonly monitorCertificateExpiration?: boolean;
}

/**
 * CloudWatch alarm actions to perform.
 */
export interface AlarmActions {
  /**
   * The ARN of the CloudWatch alarm action to take for alarms of high-severity
   * alarms.
   *
   * This must be an ARN that can be used with CloudWatch alarms.
   * @see https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/AlarmThatSendsEmail.html#alarms-and-actions
   */
  readonly highSeverity?: string;

  /**
   * The CloudWatch alarm action to take for alarms of high-severity alarms.
   *
   * This must be an ARN that can be used with CloudWatch alarms.
   * @see https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/AlarmThatSendsEmail.html#alarms-and-actions
   */
  readonly highSeverityAction?: IAlarmAction;

  /**
   * The ARN of the CloudWatch alarm action to take for alarms of normal
   * severity.
   *
   * This must be an ARN that can be used with CloudWatch alarms.
   * @see https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/AlarmThatSendsEmail.html#alarms-and-actions
   *
   * @default - no actions are taken in response to alarms of normal severity
   */
  readonly normalSeverity?: string;

  /**
   * The CloudWatch alarm action to take for alarms of normal severity.
   *
   * This must be an ARN that can be used with CloudWatch alarms.
   * @see https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/AlarmThatSendsEmail.html#alarms-and-actions
   *
   * @default - no actions are taken in response to alarms of normal severity
   */
  readonly normalSeverityAction?: IAlarmAction;
}
