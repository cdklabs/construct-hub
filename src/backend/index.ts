export * from './catalog-builder';
export * from './transliterator';
export * from './discovery';