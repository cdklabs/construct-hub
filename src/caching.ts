import { CacheControl } from '@aws-cdk/aws-s3-deployment';
import { Duration } from '@aws-cdk/core';

/**
 * Caching policies for serving data for the Construct Hub web app.
 */
export class CacheStrategy {
  /**
   * Default caching policy for all S3 objects.
   */
  public static default() {
    return new CacheStrategy([
      CacheControl.setPublic(),
      CacheControl.maxAge(Duration.minutes(5)),
      CacheControl.mustRevalidate(),
      CacheControl.sMaxAge(Duration.minutes(1)),
      CacheControl.proxyRevalidate(),
    ]);
  }

  private constructor(private readonly cacheControl: CacheControl[]) {}

  public toString() {
    return this.cacheControl.map((c) => c.value).join(', ');
  }

  public toArray() {
    return this.cacheControl;
  }
}
