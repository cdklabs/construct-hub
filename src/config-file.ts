import { mkdtempSync, writeFileSync } from 'fs';
import { tmpdir } from 'os';
import { join } from 'path';

export class ConfigFile {
  public readonly path: string;
  public readonly dir: string;
  public constructor(filename: string, content: string) {
    this.dir = mkdtempSync(join(tmpdir(), 'chconfigfile'));
    this.path = join(this.dir, filename);
    writeFileSync(this.path, content);
  }
}
