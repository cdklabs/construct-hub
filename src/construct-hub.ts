import { aws_certificatemanager, aws_route53, aws_sns } from 'aws-cdk-lib';
import { Construct } from 'constructs';

export interface ConstructHubProps {
  /**
   * The root domain name where this instance of Construct Hub will be served.
   */
  readonly hostedZone: aws_route53.IHostedZone;

  /**
   * The certificate to use for serving the Construct Hub over a custom domain.
   *
   * @default - a DNS-Validated certificate will be provisioned using the
   *            provided `hostedZone`.
   */
  readonly tlsCertificate?: aws_certificatemanager.ICertificate;

  /**
   * An optional path prefix to use for serving the Construct Hub.
   *
   * @default - none.
   */
  readonly pathPrefix?: string;

  /**
   * Contact URLs to be used for contacting this Construct Hub operators.
   *
   * @default - none
   */
  readonly contactUrls?: ContactURLs;

  /**
   * Whether the package feed from the npmjs.com registry should be enabled.
   *
   * @default true
   */
  readonly enableNpmFeed?: boolean;

  /**
   * An optional topic to be notified whenever a new package is indexed into
   * this Construct Hub instance.
   *
   * @default - none
   */
  readonly updatesTopic?: aws_sns.ITopic;

  /**
   * The name of the CloudWatch Dashboard created to observe this application.
   *
   * @default - the path to this construct is used as the dashboard name.
   */
  readonly dashboardName?: string;
}

export interface ContactURLs {
  /**
   * The URL to the issue tracker or documentation for reporting security
   * issues.
   *
   * @default - none
   */
  readonly securityIssue?: string;

  /**
   * The URL to the issue tracker or documentation for requesting a package be
   * un-listed from this Construct Hub instance.
   *
   * @default - none
   */
  readonly unlistPackage?: string;

  /**
   * The URL to the issue tracker or documentation for reporting other issues.
   *
   * @default - none
   */
  readonly other?: string;
}

export class ConstructHub extends Construct {
  public constructor(scope: Construct, id: string, _props: ConstructHubProps) {
    super(scope, id);
  }
}
