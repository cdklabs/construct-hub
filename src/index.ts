export * from './api';
export * from './construct-hub';
export * from './package-source';
export * as sources from './package-sources';
export * from './spdx-license';
export * from './package-tag';
export * from './s3/storage';
export { PackageLinkConfig, FeaturedPackages, FeaturedPackagesSection, FeaturedPackagesDetail, FeatureFlags, Category } from './webapp';
export { DomainRedirectSource } from './domain-redirect';