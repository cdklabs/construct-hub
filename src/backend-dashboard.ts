import { Dashboard, GraphWidget, GraphWidgetView, TextWidget, IWidget, MathExpression, Metric, Statistic, PeriodOverride } from '@aws-cdk/aws-cloudwatch';
import { IBucket } from '@aws-cdk/aws-s3';
import { Construct, Duration } from '@aws-cdk/core';
import { DenyList } from './backend/deny-list';
import { Ingestion } from './backend/ingestion';
import { Inventory } from './backend/inventory';
import { Orchestration } from './backend/orchestration';
import { PackageStats } from './backend/package-stats';
import { DocumentationLanguage } from './backend/shared/language';
import { ecsClusterUrl, lambdaFunctionUrl, lambdaSearchLogGroupUrl, logGroupUrl, s3ObjectUrl, sqsQueueUrl, stateMachineUrl } from './deep-link';
import { fillMetric } from './metric-utils';
import { PackageSourceBindResult } from './package-source';

export interface BackendDashboardProps {
  readonly dashboardName?: string;
  readonly packageSources: PackageSourceBindResult[];
  readonly ingestion: Ingestion;
  readonly orchestration: Orchestration;
  readonly inventory: Inventory;
  readonly denyList: DenyList;
  readonly packageData: IBucket;
  readonly packageStats?: PackageStats;
}

export class BackendDashboard extends Construct {
  public constructor(scope: Construct, id: string, props: BackendDashboardProps) {
    super(scope, id);

    new Dashboard(this, 'Resource', {
      dashboardName: props.dashboardName,
      periodOverride: PeriodOverride.AUTO,
      start: '-P1W', // Show 1 week by default
      widgets: [
        [
          new TextWidget({
            height: 2,
            width: 24,
            markdown: [
              '# Catalog Overview',
              '',
              `[button:primary:Package Data](${s3ObjectUrl(props.packageData)})`,
              `[button:Catalog Builder](${lambdaFunctionUrl(props.orchestration.catalogBuilder)})`,
              `[button:Inventory Canary](${lambdaFunctionUrl(props.inventory.function)})`,
              `[button:Search Canary Log Group](${lambdaSearchLogGroupUrl(props.inventory.function)})`,
            ].join('\n'),
          }),
        ],
        [
          new GraphWidget({
            height: 6,
            width: 12,
            title: 'Catalog Size',
            left: [
              props.inventory.metricSubmoduleCount({ label: 'Submodules' }),
              props.inventory.metricPackageVersionCount({ label: 'Package Versions' }),
              props.inventory.metricPackageMajorCount({ label: 'Package Majors' }),
              props.inventory.metricPackageCount({ label: 'Packages' }),
            ],
            leftYAxis: { min: 0 },
          }),
          new GraphWidget({
            height: 6,
            width: 12,
            title: 'Catalog Issues',
            left: [
              props.inventory.metricUnknownObjectCount({ label: 'Unknown' }),
              props.inventory.metricMissingAssemblyCount({ label: 'Missing Assembly' }),
              props.inventory.metricMissingPackageMetadataCount({ label: 'Missing Metadata' }),
              props.inventory.metricMissingPackageTarballCount({ label: 'Missing Tarball' }),
            ],
            leftYAxis: { min: 0 },
          }),
        ],
        ...this.catalogOverviewLanguageSections(props),
        ...renderPackageSourcesWidgets(props.packageSources),
        [
          new TextWidget({
            height: 2,
            width: 24,
            markdown: [
              '# Ingestion Function',
              '',
              `[button:Ingestion Function](${lambdaFunctionUrl(props.ingestion.function)})`,
              `[button:primary:Search Log Group](${lambdaSearchLogGroupUrl(props.ingestion.function)})`,
              `[button:DLQ](${sqsQueueUrl(props.ingestion.deadLetterQueue)})`,
            ].join('\n'),
          }),
        ],
        [
          new GraphWidget({
            height: 6,
            width: 12,
            title: 'Function Health',
            left: [
              fillMetric(props.ingestion.function.metricInvocations({ label: 'Invocations' })),
              fillMetric(props.ingestion.function.metricErrors({ label: 'Errors' })),
            ],
            leftYAxis: { min: 0 },
            period: Duration.minutes(1),
          }),
          new GraphWidget({
            height: 6,
            width: 12,
            title: 'Input Queue',
            left: [
              props.ingestion.queue.metricApproximateNumberOfMessagesVisible({ label: 'Visible Messages', period: Duration.minutes(1) }),
              props.ingestion.queue.metricApproximateNumberOfMessagesNotVisible({ label: 'Hidden Messages', period: Duration.minutes(1) }),
            ],
            leftYAxis: { min: 0 },
            right: [
              props.ingestion.queue.metricApproximateAgeOfOldestMessage({ label: 'Oldest Message Age', period: Duration.minutes(1) }),
            ],
            rightAnnotations: [{
              color: '#ffa500',
              label: '10 Minutes',
              value: Duration.minutes(10).toSeconds(),
            }],
            rightYAxis: { min: 0 },
            period: Duration.minutes(1),
          }),
        ],
        [
          new GraphWidget({
            height: 6,
            width: 12,
            title: 'Input Quality',
            left: [
              fillMetric(props.ingestion.metricInvalidAssembly({ label: 'Invalid Assemblies' })),
              fillMetric(props.ingestion.metricInvalidTarball({ label: 'Invalid Tarball' })),
              fillMetric(props.ingestion.metricIneligibleLicense({ label: 'Ineligible License' })),
              fillMetric(props.ingestion.metricMismatchedIdentityRejections({ label: 'Mismatched Identity' })),
              fillMetric(props.ingestion.metricFoundLicenseFile({ label: 'Found License file' })),
            ],
            leftYAxis: { label: 'Count', min: 0, showUnits: false },
            stacked: true,
          }),
          new GraphWidget({
            height: 6,
            width: 12,
            title: 'Dead Letters',
            left: [
              props.ingestion.deadLetterQueue.metricApproximateNumberOfMessagesVisible({ label: 'Visible Messages' }),
              props.ingestion.deadLetterQueue.metricApproximateNumberOfMessagesNotVisible({ label: 'Invisible Messages' }),
            ],
            leftYAxis: { min: 0 },
            right: [
              props.ingestion.deadLetterQueue.metricApproximateAgeOfOldestMessage({ label: 'Oldest Message Age' }),
            ],
            rightAnnotations: [{
              color: '#ff7f0e',
              label: '10 days',
              value: Duration.days(10).toSeconds(),
            }, {
              color: '#ff0000',
              label: '14 days (DLQ Retention)',
              value: Duration.days(14).toSeconds(),
            }],
            rightYAxis: { min: 0 },
            period: Duration.minutes(1),
          }),
        ],
        [
          new TextWidget({
            height: 2,
            width: 24,
            markdown:
              [
                '# Orchestration',
                '',
                `[button:primary:State Machine](${stateMachineUrl(props.orchestration.stateMachine)})`,
                `[button:DLQ](${sqsQueueUrl(props.orchestration.deadLetterQueue)})`,
                `[button:Redrive DLQ](${lambdaFunctionUrl(props.orchestration.redriveFunction)})`,
                `[button:Reprocess](${lambdaFunctionUrl(props.orchestration.reprocessAllFunction)})`,
              ].join('\n'),
          }),
        ],
        [
          new GraphWidget({
            height: 6,
            width: 12,
            title: 'State Machine Executions',
            left: [
              fillMetric(props.orchestration.stateMachine.metricStarted({ label: 'Started' })),
              fillMetric(props.orchestration.stateMachine.metricSucceeded({ label: 'Succeeded' })),
              fillMetric(props.orchestration.stateMachine.metricAborted({ label: 'Aborted' })),
              fillMetric(props.orchestration.stateMachine.metricFailed({ label: 'Failed' })),
              fillMetric(props.orchestration.stateMachine.metricThrottled({ label: 'Throttled' })),
              fillMetric(props.orchestration.stateMachine.metricTimedOut({ label: 'Timed Out' })),
            ],
            leftYAxis: { min: 0 },
            right: [
              props.orchestration.stateMachine.metricTime({ label: 'Duration' }),
            ],
            rightYAxis: { min: 0 },
          }),
          new GraphWidget({
            height: 6,
            width: 12,
            title: 'Dead Letter Queue',
            left: [
              props.orchestration.deadLetterQueue.metricApproximateNumberOfMessagesVisible({ label: 'Visible Messages' }),
              props.orchestration.deadLetterQueue.metricApproximateNumberOfMessagesNotVisible({ label: 'Invisible Messages' }),
            ],
            leftYAxis: { min: 0 },
            right: [
              props.orchestration.deadLetterQueue.metricApproximateAgeOfOldestMessage({ label: 'Oldest Message Age' }),
            ],
            rightAnnotations: [{
              color: '#ff7f0e',
              label: '10 days',
              value: Duration.days(10).toSeconds(),
            }, {
              color: '#ff0000',
              label: '14 days (DLQ Retention)',
              value: Duration.days(14).toSeconds(),
            }],
            rightYAxis: { min: 0 },
            period: Duration.minutes(1),
          }),
        ],

        // deny list
        // ----------------------------------------------
        [
          new TextWidget({
            height: 2,
            width: 24,
            markdown:
              [
                '# Deny List',
                '',
                `[button:primary:Deny List Object](${s3ObjectUrl(props.denyList.bucket, props.denyList.objectKey)})`,
                `[button:Prune Function](${lambdaFunctionUrl(props.denyList.prune.pruneHandler)})`,
                `[button:Prune Logs](${lambdaSearchLogGroupUrl(props.denyList.prune.pruneHandler)})`,
                `[button:Delete Queue](${sqsQueueUrl(props.denyList.prune.queue)})`,
                `[button:Delete Logs](${lambdaSearchLogGroupUrl(props.denyList.prune.deleteHandler)})`,
              ].join('\n'),
          }),
        ],
        [
          new GraphWidget({
            height: 6,
            width: 12,
            title: 'Deny List',
            left: [
              fillMetric(props.denyList.metricDenyListRules({ label: 'Rules' }), 'REPEAT'),
              props.denyList.prune.queue.metricNumberOfMessagesDeleted({ label: 'Deleted Files' }),
            ],
            leftYAxis: { min: 0 },
            period: Duration.minutes(5),
          }),
          new GraphWidget({
            height: 6,
            width: 12,
            title: 'Prune Function Health',
            left: [
              fillMetric(props.denyList.prune.pruneHandler.metricInvocations({ label: 'Invocations' })),
              fillMetric(props.denyList.prune.pruneHandler.metricErrors({ label: 'Errors' })),
            ],
            leftYAxis: { min: 0 },
            period: Duration.minutes(5),
          }),
        ],

        ...(props.packageStats ? renderPackageStatsWidgets(props.packageStats) : []),
      ],
    });
  }

  private *catalogOverviewLanguageSections({ inventory, orchestration }: BackendDashboardProps): Generator<IWidget[]> {
    yield [
      new TextWidget({
        height: 2,
        width: 24,
        markdown: [
          '# Documentation Generation',
          '',
          `[button:primary:Transliterator Logs](${logGroupUrl(orchestration.transliterator.logGroup)})`,
          `[button:Transliterator ECS Cluster](${ecsClusterUrl(orchestration.ecsCluster)})`,
        ].join('\n'),
      }),
    ];
    const mFargateUsage = new Metric({
      dimensionsMap: {
        Class: 'None',
        Resource: 'OnDemand',
        Service: 'Fargate',
        Type: 'Resource',
      },
      metricName: 'ResourceCount',
      namespace: 'AWS/Usage',
      statistic: Statistic.MAXIMUM,
    });

    yield [
      new GraphWidget({
        height: 6,
        width: 12,
        title: 'Fargate Resources',
        left: [
          mFargateUsage.with({ label: 'Fargate Usage (On-Demand)' }),
          new MathExpression({
            expression: 'SERVICE_QUOTA(mFargateUsage)',
            label: 'Fargate Quota (On-Demand)',
            usingMetrics: { mFargateUsage },
          }),
        ],
        leftYAxis: { min: 0 },
        right: [
          orchestration.metricEcsCpuUtilization({ label: 'CPU Utilization' }),
          orchestration.metricEcsMemoryUtilization({ label: 'Memory Utilization' }),
        ],
        rightYAxis: { label: 'Percent', min: 0, max: 100, showUnits: false },
      }),
      new GraphWidget({
        height: 6,
        width: 12,
        title: 'ECS Resources',
        left: [
          fillMetric(orchestration.metricEcsNetworkRxBytes({ label: 'Received Bytes' })),
          fillMetric(orchestration.metricEcsNetworkTxBytes({ label: 'Transmitted Bytes' })),
        ],
        leftYAxis: { min: 0 },
        right: [
          fillMetric(orchestration.metricEcsTaskCount({ label: 'Task Count' })),
        ],
        rightYAxis: { min: 0 },
      }),
    ];

    for (const language of DocumentationLanguage.ALL) {
      yield [
        new TextWidget({
          height: 1,
          width: 24,
          markdown: `## Language: ${language.toString()}`,
        }),
      ];
      yield [
        new GraphWidget({
          height: 6,
          width: 6,
          title: 'Package Versions',
          left: [
            inventory.metricSupportedPackageVersionCount(language, { label: 'Available', color: '#2ca02c' }),
            inventory.metricUnsupportedPackageVersionCount(language, { label: 'Unsupported', color: '#9467bd' }),
            inventory.metricMissingPackageVersionCount(language, { label: 'Missing', color: '#d62728' }),
          ],
          leftYAxis: { showUnits: false },
          view: GraphWidgetView.PIE,
        }),
        new GraphWidget({
          height: 6,
          width: 6,
          title: 'Package Versions',
          left: [
            inventory.metricSupportedPackageVersionCount(language, { label: 'Available', color: '#2ca02c' }),
            inventory.metricUnsupportedPackageVersionCount(language, { label: 'Unsupported', color: '#9467bd' }),
            inventory.metricMissingPackageVersionCount(language, { label: 'Missing', color: '#d62728' }),
          ],
          leftYAxis: { showUnits: false },
          stacked: true,
        }),
        new GraphWidget({
          height: 6,
          width: 6,
          title: 'Package Version Submodules',
          left: [
            inventory.metricSupportedSubmoduleCount(language, { label: 'Available', color: '#2ca02c' }),
            inventory.metricUnsupportedSubmoduleCount(language, { label: 'Unsupported', color: '#9467bd' }),
            inventory.metricMissingSubmoduleCount(language, { label: 'Missing', color: '#d62728' }),
          ],
          leftYAxis: { showUnits: false },
          view: GraphWidgetView.PIE,
        }),
        new GraphWidget({
          height: 6,
          width: 6,
          title: 'Package Version Submodules',
          left: [
            inventory.metricSupportedSubmoduleCount(language, { label: 'Available', color: '#2ca02c' }),
            inventory.metricUnsupportedSubmoduleCount(language, { label: 'Unsupported', color: '#9467bd' }),
            inventory.metricMissingSubmoduleCount(language, { label: 'Missing', color: '#d62728' }),
          ],
          leftYAxis: { showUnits: false },
          stacked: true,
        }),
      ];
    }
  }
}

function* renderPackageSourcesWidgets(packageSources: PackageSourceBindResult[]): Generator<IWidget[], undefined, undefined> {
  for (const packageSource of packageSources) {
    yield [
      new TextWidget({
        height: 2,
        width: 24,
        markdown: [
          `# ${packageSource.name}`,
          '',
          ...(packageSource.links ?? [])
            .map(({ name, primary, url }) => `[${primary ? 'button:primary' : 'button'}:${name}](${url})`),
        ].join('\n'),
      }),
    ];
    yield* packageSource.dashboardWidgets;
  }
  return;
}

function renderPackageStatsWidgets(packageStats: PackageStats): IWidget[][] {
  return [
    [
      new TextWidget({
        height: 2,
        width: 24,
        markdown:
          [
            '# Package Stats',
            '',
            `[button:primary:Package Stats Object](${s3ObjectUrl(packageStats.bucket, packageStats.statsKey)})`,
            `[button:Package Stats Function](${lambdaFunctionUrl(packageStats.handler)})`,
            `[button:Package Stats Logs](${lambdaSearchLogGroupUrl(packageStats.handler)})`,
          ].join('\n'),
      }),
    ],
    [
      new GraphWidget({
        height: 6,
        width: 12,
        title: 'Number of Package Stats Recorded',
        left: [
          fillMetric(packageStats.metricPackagesCount({ label: 'Packages with stats' }), 'REPEAT'),
        ],
        leftYAxis: { min: 0 },
      }),
      new GraphWidget({
        height: 6,
        width: 12,
        title: 'Invocation Duration',
        left: [
          packageStats.handler.metricDuration({ label: 'Duration' }),
        ],
        leftYAxis: { min: 0 },
        rightAnnotations: [{
          color: '#ffa500',
          label: '15 minutes (Lambda timeout)',
          value: Duration.minutes(15).toSeconds(),
        }],
      }),
    ],
  ];
}
