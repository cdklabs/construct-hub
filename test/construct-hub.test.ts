import '@aws-cdk/assert/jest';

import { App, Stack, aws_route53, cx_api } from 'aws-cdk-lib';
import { stringify as yaml } from 'yaml';

import { ConstructHub } from '../src/construct-hub';

expect.addSnapshotSerializer({
  test: (val) => val instanceof cx_api.CloudFormationStackArtifact,
  serialize: (val: cx_api.CloudFormationStackArtifact) => yaml(val.template),
});

test('minimal configuration', () => {
  // Given
  const app = new App();
  const stack = new Stack(app, 'Stack');
  const hostedZone = new aws_route53.PublicHostedZone(stack, 'HostedZone', { zoneName: 'hub.constructs.test' });

  // When
  new ConstructHub(stack, 'ConstructHub', { hostedZone });

  // Then
  expect(() => app.synth()).not.toThrow();
  expect(app.synth().getStackByName(stack.stackName)).toMatchSnapshot();
});
