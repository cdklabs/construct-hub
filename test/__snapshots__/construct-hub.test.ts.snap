// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`minimal configuration 1`] = `
Resources:
  HostedZoneDB99F866:
    Type: AWS::Route53::HostedZone
    Properties:
      Name: hub.constructs.test.
Parameters:
  BootstrapVersion:
    Type: AWS::SSM::Parameter::Value<String>
    Default: /cdk-bootstrap/hnb659fds/version
    Description: Version of the CDK Bootstrap resources in this environment,
      automatically retrieved from SSM Parameter Store.
Rules:
  CheckBootstrapVersion:
    Assertions:
      - Assert:
          Fn::Not:
            - Fn::Contains:
                - - "1"
                  - "2"
                  - "3"
                - Ref: BootstrapVersion
        AssertDescription: CDK bootstrap stack version 4 required. Please run 'cdk
          bootstrap' with a recent version of the CDK CLI.

`;
